File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/barChart/BarChartOpExec.java
package edu.uci.ics.texera.workflow.operators.visualization.barChart;

import edu.uci.ics.texera.workflow.common.operators.map.MapOpExec;
import edu.uci.ics.texera.workflow.common.tuple.Tuple;
import edu.uci.ics.texera.workflow.common.tuple.schema.Attribute;
import edu.uci.ics.texera.workflow.common.tuple.schema.OperatorSchemaInfo;
import edu.uci.ics.texera.workflow.common.tuple.schema.Schema;
import scala.Function1;
import scala.Serializable;

import java.util.ArrayList;
import java.util.List;

public class BarChartOpExec extends MapOpExec {

    private final BarChartOpDesc opDesc;
    private final OperatorSchemaInfo operatorSchemaInfo;

    public BarChartOpExec(BarChartOpDesc opDesc, OperatorSchemaInfo operatorSchemaInfo) {
        this.opDesc = opDesc;
        this.operatorSchemaInfo = operatorSchemaInfo;
        this.setMapFunc((Function1<Tuple, Tuple> & Serializable) this::processTuple);
    }

    public Tuple processTuple(Tuple t) {
        Tuple.BuilderV2 builder = Tuple.newBuilder(operatorSchemaInfo.outputSchema());
        Schema inputSchema = t.getSchema();

        builder.add(inputSchema.getAttribute(opDesc.nameColumn), t.getField(opDesc.nameColumn));

        for(String s : opDesc.dataColumns) {
            builder.add(inputSchema.getAttribute(s), t.getField(s));
        }

        return builder.build();
    }
}

File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/barChart/BarChartOpExec.scala
package edu.uci.ics.texera.workflow.operators.visualization.barChart

import edu.uci.ics.texera.workflow.common.operators.map.MapOpExec
import edu.uci.ics.texera.workflow.common.tuple.Tuple
import edu.uci.ics.texera.workflow.common.tuple.schema.OperatorSchemaInfo

/**
  * Simply mocks the data.
  * @param opDesc BarChartOpDesc.
  * @param operatorSchemaInfo The descriptor's schema info.
  */
class BarChartOpExec(
    opDesc: BarChartOpDesc,
    operatorSchemaInfo: OperatorSchemaInfo
) extends MapOpExec {

  setMapFunc(this.processTuple)

  def processTuple(t: Tuple): Tuple = {
    val builder = Tuple.newBuilder(operatorSchemaInfo.outputSchema)
    val inputSchema = t.getSchema
    builder.add(inputSchema.getAttribute(opDesc.nameColumn), t.getField(opDesc.nameColumn))
    for (i <- opDesc.resultAttributeNames.indices) {
      val dataName = opDesc.resultAttributeNames.apply(i)
      builder.add(inputSchema.getAttribute(dataName), t.getField(dataName))
    }
    builder.build
  }
}
